File: codingInterviews/求二叉树的层序遍历.java
import java.util.*;

/*
 * public class TreeNode {
 *   int val = 0;
 *   TreeNode left = null;
 *   TreeNode right = null;
 * }
 */

/**
 * 题目描述
 * 给定一个二叉树，返回该二叉树层序遍历的结果，（从左到右，一层一层地遍历）
 * 例如：
 * 给定的二叉树是{3,9,20,#,#,15,7},
 *
 * 该二叉树层序遍历的结果是
 * [
 * [3],
 * [9,20],
 * [15,7]
 * ]
 *
 * 示例1
 * 输入
 * {1,2}
 * 返回值
 * [[1],[2]]
 * 示例2
 * 输入
 * {1,2,3,4,#,#,5}
 * 返回值
 * [[1],[2,3],[4,5]]
 */

public class Solution {
    /**
     *
     * @param root TreeNode类
     * @return int整型ArrayList<ArrayList<>>
     */
    public ArrayList<ArrayList<Integer>> levelOrder (TreeNode root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (root==null){ return result;}
        Queue<TreeNode> queue= new LinkedList<TreeNode>();
        queue.offer(root);
        while(queue.size()>0){
            int len = queue.size();
            ArrayList<Integer> tmp = new ArrayList<Integer>();
            for (int i=0;i<len;i++){
                TreeNode n = queue.poll();
                tmp.add(n.val);
                if (n.left!=null){queue.offer(n.left);}
                if (n.right!=null){queue.offer(n.right);}
            }
            result.add(tmp);
        }
        return result;
    }
}

File: code-go/求二叉树的层序遍历.go
package code_go

/**
题目描述
给定一个二叉树，返回该二叉树层序遍历的结果，（从左到右，一层一层地遍历）
例如：
给定的二叉树是{3,9,20,#,#,15,7},

该二叉树层序遍历的结果是
[
[3],
[9,20],
[15,7]
]

示例1
输入
{1,2}
返回值
[[1],[2]]
示例2
输入
{1,2,3,4,#,#,5}
返回值
[[1],[2,3],[4,5]]

*/

func levelOrder(root *TreeNode) [][]int {
	result := make([][]int, 0)
	if root == nil {
		return nil
	}
	queue := make([]*TreeNode, 0)
	queue = append(queue, root)
	for len(queue) > 0 {
		size := len(queue)
		tmp := make([]int, 0)
		for i := 0; i < size; i++ {
			if len(queue) == 0 {
				break
			}
			tmp = append(tmp, queue[0].Val)
			if queue[0].Left != nil {
				queue = append(queue, queue[0].Left)
			}
			if queue[0].Right != nil {
				queue = append(queue, queue[0].Right)
			}
			queue = queue[1:]
		}
		result = append(result, tmp)
	}
	return result
}
